### PR TITLE
fix: coerce LLM string numbers to integers before phase lookups in transition phase

### PR DESCRIPTION
**Pull request by B-A-M-N**
## Problem

The `handleTransitionPhase` MCP tool expects `next_phase_id`, `completed_phase_id`, and `next_phase_ids` to be numeric types. However, LLM models can occasionally generate string numbers in function call arguments — e.g., `next_phase_id: "2"` instead of `next_phase_id: 2` — which causes strict-equality (`===`) phase lookups to fail silently.

## Fix

Adds a `coerceNumber()` helper that safely converts string numbers to actual integers before the strict-equality phase lookups run. Non-numeric strings, empty strings, floats, and non-positive values pass through unchanged so downstream errors remain accurate.

## Files Changed
- `src/mcp/handlers/session-state-tools.js` — added `coerceNumber()` and coercion logic
- `claude/src/mcp/handlers/session-state-tools.js` — generated mirror
- `plugins/maestro/src/mcp/handlers/session-state-tools.js` — generated mirror
- `tests/unit/session-state-tools.test.js` — unit tests for `coerceNumber()`

Generated files are in sync with source (zero-drift verified).

### CODE REVIEW

File: src/mcp/handlers/session-state-tools.js [hunk 1]
```diff
@@ -159,7 +159,26 @@ function handleGetSessionStatus(_params, projectRoot) {
...
+function coerceNumber(value) {
+  if (value == null) return value;
+  if (typeof value === 'number') return value;
+  const num = Number(value);
+  return Number.isNaN(num) ? value : num;
+}
+
 function handleTransitionPhase(params, projectRoot) {
+  // Coerce LLM-generated string numbers to actual numbers before schema validation
+  // LLMs often emit `next_phase_id: "2"` instead of `next_phase_id: 2`
```
Comment (josstei) [inline src/mcp/handlers/session-state-tools.js]: **Severity: Low -- documentation accuracy**

The comment says "before schema validation," but there is no runtime schema validation anywhere in the server pipeline. I traced the full call path:

1. `maestro-server.js:194-201` extracts `message.params.arguments` and passes it directly to `server.callTool(name, args, projectRoot)`
2. `create-server.js:31-42` passes `args` straight to the handler with no validation step
3. `tool-registry.js:27-72` registers `inputSchema` objects and handlers but never validates input against schemas
4. The `inputSchema` in `session/index.js:74-98` is purely informational -- it is returned to LLMs via `tools/list` but never enforced server-side
5. There is no dependency on ajv, zod, joi, or any schema validation library in `package.json`

The real issue this fix addresses is the **strict equality (`===`) comparisons** at lines 215, 226, and 237 in this file (post-patch), where `"2" === 2` evaluates to `false`.

**Suggestion:** Reword the comment to accurately describe what it prevents:

```suggestion
  // Coerce LLM-generated string numbers to actual numbers before phase-ID lookups.
  // LLMs often emit `next_phase_id: "2"` instead of `next_phase_id: 2`,
  // which breaks the strict-equality (===) comparisons below.
```

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/mcp/handlers/session-state-tools.js [hunk 2]
```diff
@@ -159,7 +159,26 @@ function handleGetSessionStatus(_params, projectRoot) {
   };
 }
 
+function coerceNumber(value) {
+  if (value == null) return value;
+  if (typeof value === 'number') return value;
+  const num = Number(value);
+  return Number.isNaN(num) ? value : num;
+}
```
Comment (josstei) [inline src/mcp/handlers/session-state-tools.js:168]: **Severity: Medium -- correctness / edge cases**

`Number()` is too permissive for phase ID coercion. Phase IDs in this codebase are always **positive integers >= 1**, established by the existing coercion in `handleCreateSession` (lines 87-89 of this file):

```js
Number.isFinite(Number(phase.id)) && Number(phase.id) > 0
    ? Number(phase.id)
    : index + 1
```

But `coerceNumber()` using bare `Number()` accepts values that can never be valid phase IDs:

| Input | `Number(input)` | Problem |
|-------|-----------------|---------|
| `""` | `0` | Empty string silently becomes 0; "Phase 0 not found" error is confusing |
| `"  "` | `0` | Whitespace silently becomes 0 |
| `true` | `1` | Boolean could falsely match phase 1 |
| `false` | `0` | Boolean silently becomes 0 |
| `"2.5"` | `2.5` | Float -- phase IDs are integers |
| `"0"` | `0` | Zero is never a valid phase ID |
| `"-1"` | `-1` | Negative is never a valid phase ID |

**Suggestion:** Align with the existing stricter pattern and rename to clarify intent:

```suggestion
function coerceNumber(value) {
  if (value == null) return value;
  if (typeof value === 'number') return value;
  const num = Number(value);
  return Number.isFinite(num) && Number.isInteger(num) && num > 0 ? num : value;
}
```

This rejects all the above edge cases and falls through to the existing error handling (`"Phase X not found"`), which will produce a clear message with the original invalid value.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/mcp/handlers/session-state-tools.js [hunk 3]
```diff
@@ -159,7 +159,26 @@ function handleGetSessionStatus(_params, projectRoot) {
...
+
 function handleTransitionPhase(params, projectRoot) {
+  // Coerce LLM-generated string numbers to actual numbers before schema validation
+  // LLMs often emit `next_phase_id: "2"` instead of `next_phase_id: 2`
+  if (params.next_phase_id != null && typeof params.next_phase_id !== 'number') {
+    params.next_phase_id = coerceNumber(params.next_phase_id);
+  }
```
Comment (josstei) [inline src/mcp/handlers/session-state-tools.js]: **Severity: Low -- code quality / redundancy**

The outer guard `params.next_phase_id != null && typeof params.next_phase_id !== 'number'` means `coerceNumber()` will never receive `null`, `undefined`, or a value that is already a `number`. But `coerceNumber()` itself already handles all three of those cases with its first two lines:

```js
if (value == null) return value;        // handles null/undefined
if (typeof value === 'number') return value;  // handles numbers
```

So the inner guards can never trigger given the outer guard. This is not a bug, but it is redundant logic that makes the code harder to reason about. The same applies to the `completed_phase_id` block on lines 175-177.

**Suggestion:** Either simplify the call sites to rely on `coerceNumber`'s built-in guards:

```js
params.next_phase_id = coerceNumber(params.next_phase_id);
params.completed_phase_id = coerceNumber(params.completed_phase_id);
```

Or, if you prefer the explicit type-check for readability, remove the redundant guards from inside `coerceNumber` and make it only handle the string-to-number conversion.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: claude/src/mcp/handlers/session-state-tools.js
Comment (josstei) [inline claude/src/mcp/handlers/session-state-tools.js:1]: **Severity: Medium -- process**

This file is a **generated mirror** of `src/mcp/handlers/session-state-tools.js`, produced by `scripts/generate.js` (invoked via `npm run generate` or `just generate`). The same applies to `plugins/maestro/src/mcp/handlers/session-state-tools.js`.

These files should not be edited directly. The correct workflow is:

1. Modify only `src/mcp/handlers/session-state-tools.js`
2. Run `npm run generate`
3. Commit the regenerated outputs

Editing generated files directly risks drift if someone later runs the generator without your changes in the source. In this case the three copies happen to be identical, so no functional issue exists today, but the process should be corrected to prevent future problems.

**Suggestion:** Remove the direct edits to this file and `plugins/maestro/src/mcp/handlers/session-state-tools.js`, then regenerate them from the canonical source.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/mcp/handlers/session-state-tools.js [hunk 4]
```diff
@@ -159,7 +159,26 @@ function handleGetSessionStatus(_params, projectRoot) {
...
+  if (params.completed_phase_id != null && typeof params.completed_phase_id !== 'number') {
+    params.completed_phase_id = coerceNumber(params.completed_phase_id);
+  }
+  if (Array.isArray(params.next_phase_ids)) {
+    params.next_phase_ids = params.next_phase_ids.map(coerceNumber);
```
Comment (josstei) [inline src/mcp/handlers/session-state-tools.js]: **Severity: Medium -- test coverage**

The new coercion logic has no test coverage. The existing test at `tests/transforms/mcp-session-pack.test.js:80-97` only exercises the happy path with integer phase IDs:

```js
completed_phase_id: 1,
next_phase_id: 2,
```

Recommended test cases to add:

1. **String IDs succeed**: `completed_phase_id: "1"` and `next_phase_id: "2"` should behave identically to their integer equivalents
2. **String array IDs succeed**: `next_phase_ids: ["1", "2"]` should coerce and start both phases
3. **Invalid strings produce clear errors**: `next_phase_id: "abc"` should fail with a "not found" error (not silently coerce)
4. **Edge cases**: `next_phase_id: ""` should fail clearly, not coerce to 0 (this also serves as a regression test for the `Number("")` edge case mentioned above)

These tests would also serve as documentation of the intended behavior for future contributors.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by josstei — CHANGES_REQUESTED** [review]
## Review Summary

Thanks for tackling this -- LLMs emitting string-typed phase IDs is a real problem, and the fix is directionally correct. The `coerceNumber` approach is sensible. I have a few concerns that should be addressed before merging.

### What works well
- Correctly identifies that string-typed IDs break `===` lookups against numeric phase IDs in session state
- Applies coercion to all three relevant parameters (`completed_phase_id`, `next_phase_id`, `next_phase_ids`)
- Non-numeric strings pass through unchanged so downstream error messages are preserved

### Requested changes

**1. Fix inaccurate PR description and comments (Low)**
The PR says "before schema validation" but there is **no runtime schema validation** in the server pipeline. `create-server.js:31` passes args directly to handlers. `tool-registry.js` registers schemas but never validates against them. The `inputSchema` is informational only (sent to LLMs via `tools/list`). The actual bug is `===` comparisons at lines 215, 226, and 237. The comment on line 170 should be corrected.

**2. Tighten `coerceNumber()` to match existing patterns (Medium)**
`Number()` accepts inputs that are never valid phase IDs: `Number("") === 0`, `Number(" ") === 0`, `Number(true) === 1`, `Number("2.5") === 2.5`. Phase IDs are always positive integers >= 1 per `handleCreateSession` (lines 87-89). The helper should use the same `Number.isFinite() && Number.isInteger() && num > 0` pattern already established in this file.

**3. Remove redundant type guards at call sites (Low)**
The `!= null && typeof !== 'number'` outer checks duplicate logic already inside `coerceNumber()`. Either simplify the call sites or the helper, not both.

**4. Only edit `src/`, regenerate mirrors (Medium)**
`claude/src/` and `plugins/maestro/src/` are generated by `scripts/generate.js`. Edit only `src/mcp/handlers/session-state-tools.js` and run `npm run generate`.

**5. Add test coverage (Medium)**
No tests exist for the string coercion path. The existing test at `tests/transforms/mcp-session-pack.test.js:80-97` uses only integer IDs. At minimum, add cases for string IDs succeeding, invalid strings failing clearly, and array coercion.

### Additional context
The MCP transport is JSON over stdio (`maestro-server.js:62-81`), so JSON natively distinguishes `2` (number) from `"2"` (string). This issue only arises when an LLM emits `"next_phase_id": "2"` as a string in its JSON output. The fix is worth having as a defensive measure, but the tighter validation will prevent silent mismatches from edge-case inputs.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by B-A-M-N** [comment]
Addressed the requested changes:

- Reworded the PR/comment to describe the actual issue (\`===\` phase-ID lookup mismatch), not schema validation
- Tightened \`coerceNumber()\` to only coerce finite positive integers — rejects \`""\`, \`" "\`, floats, zero, negatives
- Removed redundant outer type guards at call sites; \`coerceNumber\` is now idempotent
- Edited only \`src/\` and regenerated mirrors via \`npm run generate\`
- Added 13 tests covering string scalar IDs, string array IDs, invalid strings, and empty-string edge cases

All 13 new tests pass. Thanks for the detailed review.

**Review by josstei — APPROVED** [review]

**Comment by josstei** [comment]
Good to go!
